docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,1060 @@
+- Node generation ID is guaranteed to be globally unique across space and time. Node ID is not unique
+  in the event of starting a replacement node after an original node is network partitioned, or
+  in the event of a human error starting a replacement node that re-uses a node ID.
+- Node ID is written to disk in a pageserver's configuration, whereas node generation number is
+  received at runtime just after startup (this does incur an availability dependency, see [availability](#availability)).
+- The two concepts could be collapsed into one if we used ephemeral node IDs that were only ever used one time,
+  but this makes it harder to efficiently handle the case of the same physical node restarting, where we may want
+  to retain the same attachments as it had before restart, without having to map them to a different node ID.
+- We may in future run the pageserver in a different cluster runtime environment (e.g. k8s) where
+  guaranteeing that the cluster manager (e.g. k8s) doesn't run two concurrent instances with the same
+  node ID is much harder.
+
+#### Why use per-tenant _and_ per-node generation numbers?
+
+The most important generation number is the tenant attachment number: this alone would be sufficient
+to implement safe migration and failover, if we assume that our control plane will never concurrently
+run two nodes with the same node ID.
+
+Running two nodes with the same ID might sound far-fetched, but it would happen very easily if we
+ran the pageserver in a k8s environment with a StatefulSet, as k8s provides no guarantee that
+an old pod is dead before starting a new one.
+
+The node generation is useful other ways, beyond it's core correctness purpose:
+
+- Including both generations (and the node ID) in object keys also provides some flexibility in
+  how HA should work in future: we could move to a model where failover can happen within one
+  attachment generation number (i.e. without control plane coordination) without changing the
+  storage format, because node IDs/generations would de-conflict writes from peers.
+- The node generation is also useful for managing remote objects which are not per-tenant,
+  such as the persistent per-node deletion queue which is proposed in this RFC.
+- node generation numbers enable building fencing into any network protocol (for example
+  communication with safekeepers) to refuse to communicate with stale/partitioned nodes.
+
+Node that the two generation numbers have a different behavior for stale generations:
+
+- A pageserver with a stale generation number should immediately terminate: it is never intended
+  to have two pageservers with the same node ID.
+- An attachment with a stale generation number is permitted to continue operating (ingesting WAL
+  and serving reads), but not to do any deletions. This enables a multi-attached state for tenants,
+  facilitating live migration (this will be articulated in the next RFC that describes HA and
+  migrations).
```

Comment:
When reading the doc top down, the first item, esp in contrast to the second item, prompts the question:

> So, what if the pageserver doesn't immediately terminate? Are there safety implications if it does? Surely there can't be, if there were, the whole scheme would be useless.

Reply:
Added a couple of sentences here 923d4162b33b6cf0ef19a10e4a1c207bd7c6b72b